content/blog/2019-12-20-Internal-Function-Memoization/index.md

```diff
@@ -0,0 +1,175 @@
+
+Consider the call tree for recursive fibonacci.
+![](https://i.imgur.com/xaaXLDV.png)
+
+
+![](https://i.imgur.com/uCtIcSt.png)
+
+How extensible are these ideas? Consider the arbitrary one-dimensional programming problem below. Without specific knowledge of which recursive calls are required and their offsets from $n$, we must conservatively estimate a next call could use any previously computed elements.
+
+## Evaluation
+### Correctness
+In order to evaluate correctness, we look at the programs generated for the set of valid benchmarks. During construction, we separate the domain by case into separate functions, each with a different result. These _separate functions_ are the base cases, each with their own code that exeutes and returns within the case. The \{domain $\setminus$ the domain of the base cases\} is input into the loop in the body of our function. Determining program equivalence can be broken down into equivalence of these separate functions.
+
+However, function equivalence is an undecidable problem, as that would solve the [halting problem](https://en.wikipedia.org/wiki/Halting_problem). Therefore, to approximately validate our generated programs are correct, we test the bases cases, and a few arguments for the iterative case. 
+We randomly generating a few integers between 2 and 35 as 1 is a base case, and 35 is the largest fibonacci number that still fits in a 32-bit integer.
+
+A next step to improve correctness testing would be to utilize randomized testing, but instead, determine the space of valid inputs from the input program itself.
+
+### Performance
+To measure performance, we ran derivations of the fibonacci program. We compare the execution time of the recursive program to the execution time of the iterative program.
+
+Here are the results of running the fibonacci benchmark shown above. We see that the recursive benchmark follows an $O(2^n)$ execution time trendline as expected. We also see the iterative benchmark's trendline is very linear. By the data, the iterative trendline only varies between 3 and 3.5 hundredths of milliseconds between inputs 7 and 55. This is quite small. One reason this may be is that increasing n by 1 in the iterative benchmark, likely only increasing the number of add instructions by 1.
+<img src="call.png" style="max-width: 100%" >
+
+
+When interpreting these results, we considered the difference between these two programs. 
+Consider the call tree for recursive fibonacci. As $n \longrightarrow \infty$, the number of calls in the call tree grow by $O(2^n)$.
+
+<img src="fib.png" style="max-width: 100%" >
+
+In addition to the redundant computation, the recursive fibonacci program takes up more space, as it is not tail-recursive, and therefore requires many more stack frames for each computation.
+
+Next consider Katinacci, a derivation of fibonacci with the recurrence: $$k(n) = k(n-1) + k(n-3)$$
```

Comment:
I assume "Katinacci" is made up? Maybe make it clear that this is not a thing that already existed in the world.